java-checks/src/main/java/org/sonar/java/checks/SynchronizedClassUsageCheck.java

```diff
@@ -130,18 +131,34 @@ public void visitClass(ClassTree tree) {
     @Override
     public void visitMethod(MethodTree tree) {
       TypeTree returnTypeTree = tree.returnType();
-      if (isNotOverriding(tree) || returnTypeTree == null) {
+      populateOverridingMethodExclusions(tree, returnTypeTree);
+      if (!isOverriding(tree) || returnTypeTree == null) {
         if (returnTypeTree != null) {
           reportIssueOnDeprecatedType(returnTypeTree, returnTypeTree.symbolType());
         }
         scan(tree.parameters());
       }
       scan(tree.block());
+      overridingMethodTypes.pop();
     }
 
-    private boolean isNotOverriding(MethodTree tree) {
+    void populateOverridingMethodExclusions(MethodTree methodTree, TypeTree returnType) {
+      var methodTypes = new HashSet<String>();
+      if (isOverriding(methodTree)) {
+        if(returnType != null) {
+          methodTypes.add(returnType.symbolType().fullyQualifiedName());          
+        }
```

Comment:
Actually, we can probably cut some branches in this method by running some of the tests in this method in `visitMethod`.
Because we only collect types for overriding methods and we test it twice between the 2 methods.
But also because, according to the Javadoc, `returntype` can only be null for constructors and constructors can be overloaded but not overriden